qutip/tests/core/data/test_norm.py

```diff
@@ -0,0 +1,73 @@
+from . import test_mathematics as testing
+import numpy as np
+import scipy as sc
+import pytest
+from qutip import data
+from qutip.core.data import CSR, Dense
+import numbers
+
+
+class TestOneNorm(testing.UnaryOpMixin):
+    def op_numpy(self, matrix):
+        return sc.linalg.norm(matrix, 1)
+
+    specialisations = [
+        pytest.param(data.norm.one_csr, CSR, numbers.Number),
+        pytest.param(data.norm.one_dense, Dense, numbers.Number),
+    ]
+
+
+class TestFrobeniusNorm(testing.UnaryOpMixin):
+    def op_numpy(self, matrix):
+        return sc.linalg.norm(matrix, 'fro')
+
+    specialisations = [
+        pytest.param(data.norm.frobenius_csr, CSR, numbers.Number),
+        pytest.param(data.norm.frobenius_dense, Dense, numbers.Number),
+    ]
+
+
+class TestMaxNorm(testing.UnaryOpMixin):
+    def op_numpy(self, matrix):
+        # There is no scipy-equvalent as sc.linalg.norm(matrix, np.inf)
+        # works differently for matrices.
+        return np.max(np.abs(matrix))
+
+    specialisations = [
+        pytest.param(data.norm.max_csr, CSR, numbers.Number),
+        pytest.param(data.norm.max_dense, Dense, numbers.Number),
+    ]
+
+
+class TestL2Norm(testing.UnaryOpMixin):
+    def op_numpy(self, matrix):
+        return sc.linalg.norm(matrix, 'fro')
+
+    # These shapes correspond to kets or bras
+    shapes = [
+        (x,) for x in testing.shapes_unary() if (x.values[0][0] == 1
+                                                 or x.values[0][1] == 1)
+    ]
+    # These shapes are everything except for kets and bras
+    bad_shapes = [
+        (x,) for x in testing.shapes_unary() if not (x.values[0][0] == 1
+                                                     or x.values[0][1] == 1)
+    ]
+    specialisations = [
+        pytest.param(data.norm.l2_csr, CSR, numbers.Number),
+        pytest.param(data.norm.l2_dense, Dense, numbers.Number),
+    ]
+
+
+class TestTraceNorm(testing.UnaryOpMixin):
+    def op_numpy(self, matrix):
+        matrix = matrix@np.conj(matrix.T)
+        return np.trace(sc.linalg.sqrtm(matrix))
+
+    # TraceNorm probaly needs lower tol due to how it is being computed with
+    # NumPy
+    tol = 1e-5
```

Comment:
Hmmm, not keen on this.  1e-5 is really _not_ a good tolerance - we should really be passing tests better than this, I think.

Reply:
Neither do I to be honest. It seems to be pointing that something is not good enough? The change in tol is only necessary for a 2 of the tests. These are:
```
FAILED qutip/tests/core/data/test_norm.py::TestTraceNorm::test_mathematically_correct[CSR[ket,filled,sorted]->Number]
FAILED qutip/tests/core/data/test_norm.py::TestTraceNorm::test_mathematically_correct[CSR[ket,filled,unsorted]->Number]
```

Any idea what could be wrong?

Reply:
Actually, if a ket is given as input, there should be a fast path to compute the traceNorm. Will try to implement that and see if it improves.

Reply:
I'm not sure that this should even be defined for ket?  Technically the matrix norm is defined, but it doesn't make sense for a quantum object.  It's like how `l2` and `fro` are actually the same norm, but one's defined only for kets and the other's only for operators.

Reply:
At any rate, there's almost certainly a shortcut if we desperately wanted to handle it - `vector @ vector.conj().T` is a projector, whose singular values will all be calculable from the L2 norm of `vector` (probably the square of the l2 norm?), so we should be able to shortcut this calculation if we did want to handle it.

Reply:
I think the norm for kets can be computed faster by just computing the l2 norm of the ket? I will try to implement this.

Reply:
I just read the comments and I see we agree in that there is a faster shortcut. I actually think that in these two cases (fro and trace) it would be better for the data layer to handle any mathematically correct case and let qobj decide whether it makes sense to compute the norm or not. That being said, maybe we want for the moment acod the ket and bra cases with `TraceNorm` just for consistency with how `fro` handles things? We could also change `fro` to handle kets though.

Reply:
Also, 
```
FAILED qutip/tests/core/data/test_norm.py::TestTraceNorm::test_mathematically_correct[CSR[square,sparse]->Number]
```
also fails every now and then. Not sure what to do with this one apart from changing tol or repeating the test multiple times.

Reply:
The Frobenius norm already allows any shape, it's the L2 norm that rejects anything that isn't a vector.  Mathematically, the L2 norm _is_ only defined for vectors, so we should keep it like that.

For the trace norm, we need to keep the `trace_csr` function so that there's an option to pass `sparse=True` for huge systems, but for cases where `sparse=False` (the majority of cases), we can probably just change the algorithm to resolve to `scipy.linalg.norm(x, 'nuc')` - it'll be faster and more accurate.